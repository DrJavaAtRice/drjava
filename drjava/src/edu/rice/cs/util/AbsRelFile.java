/*BEGIN_COPYRIGHT_BLOCK
 *
 * Copyright (c) 2001-2017, JavaPLT group at Rice University (drjava@rice.edu)
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the names of DrJava, the JavaPLT group, Rice University, nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software is Open Source Initiative approved Open Source Software.
 * Open Source Initative Approved is a trademark of the Open Source Initiative.
 * 
 * This file is part of DrJava.  Download the current version of this project
 * from http://www.drjava.org/ or http://sourceforge.net/projects/drjava/
 * 
 * END_COPYRIGHT_BLOCK*/

package edu.rice.cs.util;

import java.io.*;
import java.net.URI;

/** A subclass of File that stores if it should be saved as absolute or relative.
  *
  * @version $Id$
  */
public class AbsRelFile extends File {
  protected boolean _keepAbsolute = false;

  public AbsRelFile(File parent, String child, boolean keepAbsolute) {
    super(parent, child);
    _keepAbsolute = keepAbsolute;
  }
  public AbsRelFile(File parent, String child) {
    this(parent, child, false);
  }
  public AbsRelFile(String pathname, boolean keepAbsolute) {
    super(pathname);
    _keepAbsolute = keepAbsolute;
  }
  public AbsRelFile(String pathname) {
    this(pathname, false);
  }
  public AbsRelFile(String parent, String child, boolean keepAbsolute) {
    super(parent, child);
    _keepAbsolute = keepAbsolute;
  }
  public AbsRelFile(String parent, String child) {
    this(parent, child, false);
  }
  public AbsRelFile(URI uri, boolean keepAbsolute) {
    super(uri);
    _keepAbsolute = keepAbsolute;
  }
  public AbsRelFile(URI uri) {
    this(uri, false);
  }
  public AbsRelFile(File f, boolean keepAbsolute) {
    this(f.getParent(), f.getName(), keepAbsolute);
  }
  public AbsRelFile(File f) {
    this(f, false);
  }
  public boolean keepAbsolute() { return _keepAbsolute; }
  public AbsRelFile keepAbsolute(boolean keepAbsolute) {
    _keepAbsolute = keepAbsolute;
    return this;
  }
}
