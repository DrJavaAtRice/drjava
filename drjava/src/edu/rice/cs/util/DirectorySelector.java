/*BEGIN_COPYRIGHT_BLOCK
 *
 * Copyright (c) 2001-2017, JavaPLT group at Rice University (drjava@rice.edu)
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the names of DrJava, the JavaPLT group, Rice University, nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software is Open Source Initiative approved Open Source Software.
 * Open Source Initative Approved is a trademark of the Open Source Initiative.
 * 
 * This file is part of DrJava.  Download the current version of this project
 * from http://www.drjava.org/ or http://sourceforge.net/projects/drjava/
 * 
 * END_COPYRIGHT_BLOCK*/

package edu.rice.cs.util;

import java.io.File;
/** An interface to give GlobalModel a directory for various operations.
  *
  * @version $Id$
  */
public interface DirectorySelector {
  
  /** Gets a directory.
    * @param start a file to start the search from, or null to allow implementation to determine
    * @return any valid directory on the file system
    * @throws OperationCanceledException if the selection request is canceled
    */
  public File getDirectory(File start) throws OperationCanceledException;
  
  /** Asks the user a yes/no question.
   * @param message message to be displayed
   * @param title title 
   * @return true if the user responded affirmatively, false if negatively
   */
  public boolean askUser(String message, String title);
  
  /** Warns the user about an error condition. 
   * @param message message to be displayed
   * @param title title 
   */
  public void warnUser(String message, String title);
  
  /** @return true if the user wants the directory to include all subdirectories. */
  public boolean isRecursive();
  
}
