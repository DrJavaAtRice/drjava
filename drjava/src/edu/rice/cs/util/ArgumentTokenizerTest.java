/*BEGIN_COPYRIGHT_BLOCK
 *
 * Copyright (c) 2001-2017, JavaPLT group at Rice University (drjava@rice.edu)
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the names of DrJava, the JavaPLT group, Rice University, nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software is Open Source Initiative approved Open Source Software.
 * Open Source Initative Approved is a trademark of the Open Source Initiative.
 * 
 * This file is part of DrJava.  Download the current version of this project
 * from http://www.drjava.org/ or http://sourceforge.net/projects/drjava/
 * 
 * END_COPYRIGHT_BLOCK*/

package edu.rice.cs.util;

import edu.rice.cs.drjava.DrJavaTestCase;

import java.util.Arrays;
import java.util.List;

/** * Tests that an ArgumentTokenizer can correctly divide up a string
 * into command line-style arguments.  Tries to follow the precedent
 * set by a Unix bash shell in most cases.
 * @version $Id$
 */
public class ArgumentTokenizerTest extends DrJavaTestCase {

  /** Creates a new ArgumentTokenizer to be used in every test.
   * @param name test name
   */
  public ArgumentTokenizerTest(String name) {
    super(name);
  }

  /** Asserts that the given string is tokenized to become the
   * given array of string arguments.
   * @param typed A string containing all arguments (as typed by a user)
   * @param expected What the tokenizer should return
   */
  protected void _assertTokenized(String typed, String[] expected) {
    _assertTokenized(typed, expected, false);
  }

  /** Asserts that the given string is tokenized to become the
   * given array of string arguments.
   * @param typed A string containing all arguments (as typed by a user)
   * @param expected What the tokenizer should return
   * @param stringify Whether to format the resulting arguments to print
   * out as Strings.
   */
  protected void _assertTokenized(String typed, String[] expected,
                                  boolean stringify) {
    List<String> actual = ArgumentTokenizer.tokenize(typed, stringify);
    List<String> expectedList = Arrays.asList(expected);
    assertEquals("tokenized argument list should match expected",
                 expectedList, actual);
  }

  /** Tests that the argument tokenizer can split up a simple list of arguments.
   */
  public void testTokenizeArguments() {
    // a b c
    // [a, b, c]
    _assertTokenized("a b c",
                     new String[]{"a","b","c"});
    // "a b c"
    // [a b c]
    _assertTokenized("\"a b c\"",
                     new String[]{"a b c"});

    // "a b"c d
    // [a bc, d]
    // This behavior seems unintuitive, but it's the way both DOS and Unix
    //  handle command-line arguments.
    _assertTokenized("\"a b\"c d",
                     new String[]{"a bc","d"});

    // 'a b'c d
    // [a bc, d]
    // This behavior seems unintuitive, but it's the way both DOS and Unix
    //  handle command-line arguments.
    _assertTokenized("'a b'c d",
                     new String[]{"a bc","d"});

    // a b"c d"
    // [a, bc d]
    // This behavior seems unintuitive, but it's the way both DOS and Unix
    //  handle command-line arguments.
    _assertTokenized("a b\"c d\"",
                     new String[]{"a","bc d"});

    // a b'c d'
    // [a, bc d]
    // This behavior seems unintuitive, but it's the way both DOS and Unix
    //  handle command-line arguments.
    _assertTokenized("a b'c d'",
                     new String[]{"a","bc d"});

    // a b'c d'"e f" g "hi "
    // [a, bc de f, g, hi ]
    _assertTokenized("a b'c d'\"e f\" g \"hi \"",
                     new String[]{"a","bc de f","g","hi "});

    // c:\\file.txt
    // [c:\file.txt]
    _assertTokenized("c:\\\\file.txt",
                     new String[]{"c:\\file.txt"});

    // /home/user/file
    // [/home/user/file]
    _assertTokenized("/home/user/file",
                     new String[]{"/home/user/file"});

    // "asdf
    // [asdf]
    _assertTokenized("\"asdf",
                     new String[]{"asdf"});
  }

  /** Tests that escaped characters just return the character itself.
   * Escaped whitespace is considered a character, not a delimiter.
   * (This is how Unix behaves.)
   *
   * not currently enforcing any behavior for a simple implementation
   * using a StreamTokenizer
   */
  public void testTokenizeEscapedArgs() {
    // \j
    // [j]
    _assertTokenized("\\j",
                     new String[]{"j"});
    // \"
    // ["]
    _assertTokenized("\\\"",
                     new String[]{"\""});
    // \\
    // [\]
    _assertTokenized("\\\\",
                     new String[]{"\\"});
    // a\ b
    // [a b]
    _assertTokenized("a\\ b",
                     new String[]{"a b"});
  }

  /** Tests that within a quote, everything is correctly escaped.
   * (Special characters are passed to the program correctly.)
   */
  public void testTokenizeQuotedEscapedArgs() {
    // "a \" b"
    // [a " b]
    _assertTokenized("\"a \\\" b\"",
                     new String[]{"a \" b"});
    // "\'"
    // [\']
    _assertTokenized("\"'\"",
                     new String[]{"'"});
    // "\\"
    // [\]
    _assertTokenized("\\\\",
                     new String[]{"\\"});
    // "\" \d"
    // [" \d]
    _assertTokenized("\"\\\" \\d\"",
                     new String[]{"\" \\d"});
    // "\n"
    // [\n]
    _assertTokenized("\"\\n\"",
                     new String[]{"\\n"});
    // "\t"
    // [\t]
    _assertTokenized("\"\\t\"",
                     new String[]{"\\t"});
    // "\r"
    // [\r]
    _assertTokenized("\"\\r\"",
                     new String[]{"\\r"});
    // "\f"
    // [\f]
    _assertTokenized("\"\\f\"",
                     new String[]{"\\f"});
    // "\b"
    // [\b]
    _assertTokenized("\"\\b\"",
                     new String[]{"\\b"});
  }

  /** Tests that single quotes can be used as argument delimiters.
   * This is consistent with Unix, not with DOS.
   */
  public void testTokenizeSingleQuotedArgs() {
    // 'asdf'
    // [asdf]
    _assertTokenized("'asdf'",
                     new String[]{"asdf"});
    // 'a b c'
    // [a b c]
    _assertTokenized("'a b c'",
                     new String[]{"a b c"});
    // '\'
    // [\]
    _assertTokenized("'\\'",
                     new String[]{"\\"});
  }

  /** Tests that arguments can be "stringified" properly.
   * (ie. formatted to be printed as a String)
   */
  public void testTokenizeAndStringify() {
    // a b c
    // ["a", "b", "c"]
    _assertTokenized("a b c",
                     new String[]{"\"a\"", "\"b\"", "\"c\""},
                     true);
    // \\
    // ["\\"]
    _assertTokenized("\\",
                     new String[]{"\"\\\\\""},
                     true);
    // \"
    // ["\""]
    _assertTokenized("\\\"",
                     new String[]{"\"\\\"\""},
                     true);
    // "\n"
    // ["\\n"]
    _assertTokenized("\"\\n\"",
                     new String[]{"\"\\\\n\""},
                     true);
  }
}
