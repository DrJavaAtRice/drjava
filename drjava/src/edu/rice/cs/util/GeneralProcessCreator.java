/*BEGIN_COPYRIGHT_BLOCK
 *
 * Copyright (c) 2001-2017, JavaPLT group at Rice University (drjava@rice.edu)
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the names of DrJava, the JavaPLT group, Rice University, nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software is Open Source Initiative approved Open Source Software.
 * Open Source Initative Approved is a trademark of the Open Source Initiative.
 * 
 * This file is part of DrJava.  Download the current version of this project
 * from http://www.drjava.org/ or http://sourceforge.net/projects/drjava/
 * 
 * END_COPYRIGHT_BLOCK*/

package edu.rice.cs.util;

import edu.rice.cs.drjava.config.PropertyMaps;
import edu.rice.cs.plt.iter.IterUtil;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** * This class abstracts out process creation, similar to ProcessCreator,
 * which is only available in Java 1.5. Additionally, it transparently
 * creates process sequences and process chains, including piping.
 */

public class GeneralProcessCreator extends ProcessCreator {
  protected volatile List<List<List<String>>> _seqs;
    
  /** Constructor for a process creator with the given command line and the work directory.
    * @param cmdline command line
    * @param workdir working directory
    * @param pm PropertyMaps object used for substitution
    */
  public GeneralProcessCreator(String cmdline, String workdir, PropertyMaps pm) {
    _cmdline = cmdline;
    _workdir = workdir;
    _props = pm;
  }

  /** Constructor for a process creator with the given command line already split up, and
    * the work directory.
    * @param seqs a sequence of commands to pipe
    * @param workdir working directory
    * @param pm PropertyMaps object used for substitution
    */
  public GeneralProcessCreator(List<List<List<String>>> seqs, String workdir, PropertyMaps pm) {
    _seqs = seqs;
    _workdir = workdir;
    _props = pm;
  }
  
  /** Reconstructs the command line for a simple process. 
   * @param cmds commands to be reconstructed
   * @return newly-constructed command line
   */
  protected static String getProcessCmdLine(List<String> cmds) {
    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < cmds.size(); ++i) {
      sb.append(" ");
      sb.append(StringOps.unescapeFileName(cmds.get(i)));
    }
    String s = sb.toString();
    if (s.length() > 0) {
      s = s.substring(1);
    }
    return s;
  }
  
  /** Reconstructs the command line for a process chain. 
   * @param pipe pipe from which to reconstruct the command line
   * @return newly-constructed command line
   */
  protected static String getProcessChainCmdLine(List<List<String>> pipe) {
    StringBuilder sb = new StringBuilder();
    final String sep = " " + ProcessChain.PIPE_SEPARATOR + " ";
    for (int i = 0; i < pipe.size(); ++i) {
      sb.append(sep);
      sb.append(getProcessCmdLine(pipe.get(i)));
    }
    String s = sb.toString();
    if (s.length() > 0) {
      s = s.substring(sep.length());
    }
    return s;
  }
  
  /** Reconstructs the command line for a process sequence. 
   * @param seqs sequences from which to reconstruct the command line
   * @return newly-constructed command line
   */
  protected static String getProcessSequenceCmdLine(List<List<List<String>>> seqs) {
    StringBuilder sb = new StringBuilder();
    final String sep = " " + ProcessChain.PROCESS_SEPARATOR + " ";
    for (int i = 0; i < seqs.size(); ++i) {
      sb.append(sep);
      sb.append(getProcessChainCmdLine(seqs.get(i)));
    }
    String s = sb.toString();
    if (s.length() > 0) {
      s = s.substring(sep.length());
    }
    return s;
  }
  
  /** Get the command line.
    * @return command line
    */
  public String cmdline() {
    if (_cmdline == null) {
      if (_cachedCmdLine == null) {
        if (_seqs.size() == 1) {
          // only one piping chain, creating a process sequence is not necessary
          List<List<String>> pipe = _seqs.get(0);
          if (pipe.size() == 1) {
            // only one process, creating a piping chain is not necessary
            List<String> cmds = pipe.get(0);
            _cachedCmdLine = getProcessCmdLine(cmds);
          }
          else {
            // more than one process, create a process chain
            _cachedCmdLine = getProcessChainCmdLine(pipe);
          }
        }
        else  {
          // more than one piping chain, create a process sequence
          _cachedCmdLine = getProcessSequenceCmdLine(_seqs);
        }
      }
      return _cachedCmdLine;
    }
    else {
      return _cmdline;
    }
  }
  
  /** Returns a map of this process creator's environment.
    * @return environment map
    */
  public Map<String,String> environment() {
    return _env;
  }
  
  /** Returns this process creator's working directory.
    * @return working directory
    */
  public String workDir() {
    return _workdir;
  }

  public static final edu.rice.cs.util.Log LOG = new edu.rice.cs.util.Log("process.txt",false);
  
  /** Starts a new process using the attributes of this process creator. */
  public Process start() throws IOException {
    // set up work directory
    _evaluatedWorkDir = StringOps.replaceVariables(_workdir, _props, PropertyMaps.GET_CURRENT);
    _evaluatedWorkDir = StringOps.unescapeFileName(_evaluatedWorkDir);
    File dir = null;
    if (!_evaluatedWorkDir.trim().equals("")) { dir = new File(_evaluatedWorkDir); }

    // set up environment
    String[] env = null;
    if ((_env != null) && (_env.size() > 0)) {
      env = new String[_env.size()];
      int i = 0;
      for(String key: _env.keySet()) {
        String value = _env.get(key);
        env[i] = key + "=" + value;
      }
    }

    // set up command line
    if (_cmdline != null) {
      _evaluatedCmdLine = StringOps.replaceVariables(_cmdline, _props, PropertyMaps.GET_CURRENT);
      _seqs = StringOps.commandLineToLists(_evaluatedCmdLine);
    }
    LOG.log("\t" + IterUtil.toString(_seqs));
    if (_seqs.size()<1) { throw new IOException("No process to start."); }
    if (_seqs.size() == 1) {
      // only one piping chain, creating a process sequence is not necessary
      List<List<String>> pipe = _seqs.get(0);
      if (pipe.size()<1) { throw new IOException("No process to start."); }
      if (pipe.size() == 1) {
        // only one process, creating a piping chain is not necessary
        List<String> cmds = pipe.get(0);
        if (cmds.size()<1) { throw new IOException("No process to start."); }
        String[] cmdarray = new String[cmds.size()];
        for (int i = 0; i < cmds.size(); ++i) {
          cmdarray[i] = StringOps.unescapeFileName(cmds.get(i));
        }
        // creating a simple process
        return Runtime.getRuntime().exec(cmdarray,env,dir);
      }
      // more than one process, create a process chain
      ProcessCreator[] creators = new ProcessCreator[pipe.size()];
      for (int i = 0; i < pipe.size(); ++i) {
        List<String> cmds = pipe.get(i);
        if (cmds.size()<1) { throw new IOException("No process to start."); }
        String[] cmdarray = new String[cmds.size()];
        for (int j=0; j<cmds.size(); ++j) {
          cmdarray[j] = StringOps.unescapeFileName(cmds.get(j));
        }
        creators[i] = new ProcessCreator(cmdarray, _workdir);
      }
      return new ProcessChain(creators);
    }
    // more than one piping chain, create a process sequence
    ProcessCreator[] creators = new ProcessCreator[_seqs.size()];
    for (int i = 0; i < _seqs.size(); ++i) {
      List<List<List<String>>> l = new ArrayList<List<List<String>>>();
      l.add(_seqs.get(i));
      creators[i] = new GeneralProcessCreator(l, _workdir, _props);
    }
    return new ProcessSequence(creators);
  }
}
