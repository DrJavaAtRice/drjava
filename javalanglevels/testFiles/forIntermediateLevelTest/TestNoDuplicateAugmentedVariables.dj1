import java.util.*;

class TestNoDuplicateAugmentedVariables {
  LinkedList __hashCodeList;
  LinkedList __hashCodeList0;
  LinkedList __equalsList;
  boolean __toStringFlag;
  boolean __toStringFlag0;
  boolean __toStringFlag2;
  
}