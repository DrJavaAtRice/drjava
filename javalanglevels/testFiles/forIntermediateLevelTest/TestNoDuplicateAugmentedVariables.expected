// Language Level Converter line number map: dj*->java. Entries: 11
//     1->4         2->5         3->6         4->7         5->8         6->9         7->10        8->11   
//     9->12       10->13       11->14   
import java.util.*;

class TestNoDuplicateAugmentedVariables {
  private final LinkedList __hashCodeList;
  private final LinkedList __hashCodeList0;
  private final LinkedList __equalsList;
  private final boolean __toStringFlag;
  private final boolean __toStringFlag0;
  private final boolean __toStringFlag2;
  

  /** This method is automatically generated by the Language Level Converter. */
  public TestNoDuplicateAugmentedVariables(java.util.LinkedList __hashCodeList, java.util.LinkedList __hashCodeList0, java.util.LinkedList __equalsList, boolean __toStringFlag, boolean __toStringFlag0, boolean __toStringFlag2) {
    super();
    this.__hashCodeList = __hashCodeList;
    this.__hashCodeList0 = __hashCodeList0;
    this.__equalsList = __equalsList;
    this.__toStringFlag = __toStringFlag;
    this.__toStringFlag0 = __toStringFlag0;
    this.__toStringFlag2 = __toStringFlag2;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public java.util.LinkedList __hashCodeList() {
    return __hashCodeList;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public java.util.LinkedList __hashCodeList0() {
    return __hashCodeList0;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public java.util.LinkedList __equalsList() {
    return __equalsList;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public boolean __toStringFlag() {
    return __toStringFlag;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public boolean __toStringFlag0() {
    return __toStringFlag0;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public boolean __toStringFlag2() {
    return __toStringFlag2;
  }

  /** This method is automatically generated by the Language Level Converter. */
  public java.lang.String toString() {
    return getClass().getName() + "(" + 
        __hashCodeList() + ", " + 
        __hashCodeList0() + ", " + 
        __equalsList() + ", " + 
        __toStringFlag() + ", " + 
        __toStringFlag0() + ", " + 
        __toStringFlag2() + 
        ")";
  }

  /** This method is automatically generated by the Language Level Converter. */
  public boolean equals(java.lang.Object o) {
    if (this == o) {
      return true;
    }
    else if ((o == null) || (! o.getClass().equals(getClass()))) {
      return false;
    }
    else {
      TestNoDuplicateAugmentedVariables cast = ((TestNoDuplicateAugmentedVariables) o);
      return (__hashCodeList() != null && __hashCodeList().equals(cast.__hashCodeList())) && 
          (__hashCodeList0() != null && __hashCodeList0().equals(cast.__hashCodeList0())) && 
          (__equalsList() != null && __equalsList().equals(cast.__equalsList())) && 
          (__toStringFlag() == cast.__toStringFlag()) && 
          (__toStringFlag0() == cast.__toStringFlag0()) && 
          (__toStringFlag2() == cast.__toStringFlag2());
    }
  }

  /** This method is automatically generated by the Language Level Converter. */
  public int hashCode() {
    return getClass().hashCode() ^ 
        (__hashCodeList() == null ? 0 : __hashCodeList().hashCode()) ^ 
        (__hashCodeList0() == null ? 0 : __hashCodeList0().hashCode()) ^ 
        (__equalsList() == null ? 0 : __equalsList().hashCode()) ^ 
        (__toStringFlag() ? 1 : 0) ^ 
        (__toStringFlag0() ? 1 : 0) ^ 
        (__toStringFlag2() ? 1 : 0);
  }
}
